File: sx126x.py
import RPi.GPIO as GPIO
import serial
import time
import datetime

class sx126x:

    # Define operating modes (constants)
    MODE_STDBY = 0x01
    MODE_TX = 0x02
    MODE_RX = 0x03
    MODE_SLEEP = 0x04

    # Raspberry Pi GPIO pins connected to M0 and M1
    M0 = 22
    M1 = 27

    # Default configuration register values
    DEFAULT_CONFIG = [
        0xC2,  # Header
        0x00,  # MSB of address (Placeholder)
        0x09,  # LSB of address (Placeholder)
        0x00,  # Net ID
        0x00,  # High byte of own address
        0x00,  # Low byte of own address
        0x62,  # UART baud rate (9600) + Air Data Rate (2.4k)
        0x17,  # Packet size (240 bytes) + Power (22dBm) + WOR Disable
        0x00,  # Channel
        0x03,  # Options: RSSI byte + TX en
        0x00,  # MSB of encryption key
        0x00   # LSB of encryption key
    ]

    def __init__(self, serial_num, freq=433, addr=0, power=22, rssi=False):
        self.serial_n = serial_num
        self.freq = freq
        self.addr = addr  # *Own* address of THIS module
        self.power = power
        self.rssi = rssi
        self.modem = None

        GPIO.setmode(GPIO.BCM)
        GPIO.setwarnings(False)
        GPIO.setup(self.M0, GPIO.OUT)
        GPIO.setup(self.M1, GPIO.OUT)
        self.set_mode(self.MODE_STDBY)

        try:
            self.ser = serial.Serial(serial_num, 9600)
            self.flush() # Initial flush
        except serial.SerialException as e:
            print(f"Error opening serial port: {e}")
            raise

        self.set(freq, addr, power, rssi)

    def set_mode(self, mode):
        if mode == self.MODE_RX:
            GPIO.output(self.M0, GPIO.LOW)
            GPIO.output(self.M1, GPIO.LOW)
        elif mode == self.MODE_TX:
            GPIO.output(self.M0, GPIO.HIGH)
            GPIO.output(self.M1, GPIO.LOW)
        elif mode == self.MODE_STDBY:
            GPIO.output(self.M0, GPIO.LOW)
            GPIO.output(self.M1, GPIO.HIGH)
        elif mode == self.MODE_SLEEP:
            GPIO.output(self.M0, GPIO.HIGH)
            GPIO.output(self.M1, GPIO.HIGH)
        else:
            raise ValueError("Invalid mode")
        self.modem = mode
        time.sleep(0.02)

    def write_payload(self, payload):
        self.ser.write(bytes(payload))

    def read_payload(self, length, timeout=1):
        self.ser.timeout = timeout
        data = self.ser.read(length)
        self.ser.timeout = None
        return data

    def set(self, freq, addr, power, rssi, air_speed=2400, net_id=0, buffer_size=240, crypt=0):
        self.addr = addr
        config = self.DEFAULT_CONFIG[:]
        config[3] = net_id & 0xFF
        config[4] = (addr >> 8) & 0xFF
        config[5] = addr & 0xFF

        if freq >= 850:
            freq_reg = freq - 850
        elif freq >= 410:
            freq_reg = freq - 410
        else:
            raise ValueError("Invalid frequency")
        config[8] = freq_reg

        air_speed_setting = {
            300:   0x00,
            1200:  0x01,
            2400:  0x02,
            4800:  0x03,
            9600:  0x04,
            19200: 0x05,
            38400: 0x06,
            62500: 0x07,
        }.get(air_speed, 0x02)
        config[6] = 0x60 | air_speed_setting

        buffer_size_setting = {
            240: 0x00,
            128: 0x40,
            64:  0x80,
            32:  0xC0,
        }.get(buffer_size, 0x00)
        power_setting = {
            22: 0x00,
            17: 0x01,
            13: 0x02,
            10: 0x03,
        }.get(power, 0x00)
        config[7] = buffer_size_setting | power_setting

        if rssi:
            config[9] |= 0x80
        else:
            config[9] &= ~0x80

        config[10] = (crypt >> 8) & 0xFF
        config[11] = crypt & 0xFF

        self.set_mode(self.MODE_STDBY)
        self.flush()
        self.write_payload([0xC2, 0x00, 0x09] + config[3:])

        response = self.read_payload(3)
        if not (response and response[0] == 0xC1):
            print(f"Configuration failed. Response: {response.hex() if response else 'No response'}")
        self.set_mode(self.MODE_RX)

    def send(self, destination_address, data):
        if isinstance(data, str):
            data = data.encode('utf-8')
        if not isinstance(data, bytes):
            raise TypeError("Data must be bytes or a string")

        packet = [(destination_address >> 8) & 0xFF, destination_address & 0xFF] + list(data)
        print(f"Sending packet: {bytes(packet).hex()}")
        self.set_mode(self.MODE_TX)
        self.write_payload(packet)
        time.sleep(0.1)
        self.set_mode(self.MODE_RX)
        self.flush()


    def receive(self, timeout=5):
        self.set_mode(self.MODE_RX)
        self.flush()
        start_time = time.time()
        received_data = bytearray()

        while time.time() - start_time < timeout:
            if self.ser.inWaiting() > 0:
                print(f"inWaiting: {self.ser.inWaiting()}")  # DEBUG: How many bytes?
                received_data += self.ser.read(self.ser.inWaiting())
                print(f"Received raw bytes: {received_data.hex()}")

                if len(received_data) >= 2:
                    destination_address = (received_data[0] << 8) | received_data[1]
                    print(f"Received message intended for address {destination_address}")

                    if destination_address == self.addr or destination_address == 65535:
                        if len(received_data) >= 4:
                            sender_address = (received_data[2] << 8) | received_data[3]
                            print(f"Sender address: {sender_address}")
                            payload = received_data[4:]

                            if self.rssi and len(payload) > 0:
                                rssi_value = -(256 - payload[-1])
                                print(f"RSSI: {rssi_value} dBm")
                                payload = payload[:-1]

                            return payload
                        else:
                            print("Incomplete message (no sender address)")
                            received_data = bytearray()
                            return None
                    else:
                        print(f"Message not for us (destination: {destination_address}, our address: {self.addr})")
                        received_data = bytearray()
                        return None
            time.sleep(0.01)

        return None

    def cancel_receive(self):
        self.set_mode(self.MODE_STDBY)
        self.flush()

    def flush(self):
        self.ser.reset_input_buffer()

File: sx126x.js
class SX126X {
    constructor(serial_num, freq, addr, power, rssi) { // Removed serial_num from constructor args
        this.rssi = rssi;
        this.addr = addr;
        this.freq = freq;
        this.serial_n = serial_num; // You can keep this if you need to store serial_num, but it's not used for init here
        this.power = power;
        this.send_to = addr;

        this.M0 = 22;
        this.M1 = 27;

        this.cfg_reg = [0xC2, 0x00, 0x09, 0x00, 0x00, 0x00, 0x62, 0x00, 0x17, 0x00, 0x00, 0x00];
        this.get_reg = Buffer.alloc(12);
        this.serialPort = null;

        this.SX126X_UART_BAUDRATE_1200 = 0x00;
        this.SX126X_UART_BAUDRATE_2400 = 0x20;
        this.SX126X_UART_BAUDRATE_4800 = 0x40;
        this.SX126X_UART_BAUDRATE_9600 = 0x60;
        this.SX126X_UART_BAUDRATE_19200 = 0x80;
        this.SX126X_UART_BAUDRATE_38400 = 0xA0;
        this.SX126X_UART_BAUDRATE_57600 = 0xC0;
        this.SX126X_UART_BAUDRATE_115200 = 0xE0;

        this.SX126X_AIR_SPEED_300bps = 0x00;
        this.SX126X_AIR_SPEED_1200bps = 0x01;
        this.SX126X_AIR_SPEED_2400bps = 0x02;
        this.SX126X_AIR_SPEED_4800bps = 0x03;
        this.SX126X_AIR_SPEED_9600bps = 0x04;
        this.SX126X_AIR_SPEED_19200bps = 0x05;
        this.SX126X_AIR_SPEED_38400bps = 0x06;
        this.SX126X_AIR_SPEED_62500bps = 0x07;

        this.SX126X_PACKAGE_SIZE_240_BYTE = 0x00;
        this.SX126X_PACKAGE_SIZE_128_BYTE = 0x40;
        this.SX126X_PACKAGE_SIZE_64_BYTE = 0x80;
        this.SX126X_PACKAGE_SIZE_32_BYTE = 0xC0;

        this.SX126X_Power_22dBm = 0x00;
        this.SX126X_Power_17dBm = 0x01;
        this.SX126X_Power_13dBm = 0x02;
        this.SX126X_Power_10dBm = 0x03;

        this.initializeGPIO();
        // this.initializeSerial(serial_num);  <-- REMOVED from constructor
        this.set(freq, addr, power, rssi);
    }

    beginSerial(serial_num) { // Renamed method
        // const SerialPort = require('serialport');
        const SerialPort = require('serialport').SerialPort;

        this.serialPort = new SerialPort({ // e.g., '/dev/ttyS0'
            path: serial_num,
            baudRate: 9600,
        });

        this.serialPort.on('open', () => {
            this.serialPort.flush(() => {
                console.log('Serial port opened and input flushed');
            });
        });

        this.serialPort.on('error', (err) => {
            console.error("Serial port error:", err);
        });

    }

    async initializeGPIO() {
        const pigpio = require('pigpio').Gpio; // Import pigpio library
        this.m0Pin = new pigpio(this.M0, { mode: pigpio.OUTPUT });
        this.m1Pin = new pigpio(this.M1, { mode: pigpio.OUTPUT });

        this.m0Pin.digitalWrite(0); // GPIO.LOW in pigpio (0)
        this.m1Pin.digitalWrite(1); // GPIO.HIGH in pigpio (1)
    }

    initializeSerial(serial_num) {
        // const SerialPort = require('serialport');
        const SerialPort = require('serialport').SerialPort;

        this.serialPort = new SerialPort(  { // e.g., '/dev/ttyS0'
            path: serial_num,
            baudRate: 9600,
        });

        this.serialPort.on('open', () => {
            this.serialPort.flush(() => { // Flush input buffer on open
                console.log('Serial port opened and input flushed');
            });
        });

        this.serialPort.on('error', (err) => {
            console.error("Serial port error:", err);
        });

    }


    air_speed_cal(air_speed) {
        // TODO: Implement air speed calculation logic (refer to Python code if available)
        // This is a placeholder, you'll need to create the conversion logic
        if (air_speed === 300) return this.SX126X_AIR_SPEED_300bps;
        else if (air_speed === 1200) return this.SX126X_AIR_SPEED_1200bps;
        else if (air_speed === 2400) return this.SX126X_AIR_SPEED_2400bps;
        else if (air_speed === 4800) return this.SX126X_AIR_SPEED_4800bps;
        else if (air_speed === 9600) return this.SX126X_AIR_SPEED_9600bps;
        else if (air_speed === 19200) return this.SX126X_AIR_SPEED_19200bps;
        else if (air_speed === 38400) return this.SX126X_AIR_SPEED_38400bps;
        else if (air_speed === 62500) return this.SX126X_AIR_SPEED_62500bps;
        else return this.SX126X_AIR_SPEED_2400bps; // Default to 2400 if not matched
    }

    buffer_size_cal(buffer_size) {
        // TODO: Implement buffer size calculation logic
        if (buffer_size === 240) return this.SX126X_PACKAGE_SIZE_240_BYTE;
        else if (buffer_size === 128) return this.SX126X_PACKAGE_SIZE_128_BYTE;
        else if (buffer_size === 64) return this.SX126X_PACKAGE_SIZE_64_BYTE;
        else if (buffer_size === 32) return this.SX126X_PACKAGE_SIZE_32_BYTE;
        else return this.SX126X_PACKAGE_SIZE_240_BYTE; // Default to 240
    }

    power_cal(power) {
        // TODO: Implement power calculation logic
        if (power === 22) return this.SX126X_Power_22dBm;
        else if (power === 17) return this.SX126X_Power_17dBm;
        else if (power === 13) return this.SX126X_Power_13dBm;
        else if (power === 10) return this.SX126X_Power_10dBm;
        else return this.SX126X_Power_22dBm; // Default to 22dBm
    }


    async set(freq, addr, power, rssi, air_speed = 2400,
        net_id = 0, buffer_size = 240, crypt = 0,
        relay = false, lbt = false, wor = false) {
        this.send_to = addr;
        this.addr = addr;

        this.m0Pin.digitalWrite(0); // GPIO.LOW
        this.m1Pin.digitalWrite(1); // GPIO.HIGH
        await new Promise(resolve => setTimeout(resolve, 100)); // time.sleep(0.1)

        let low_addr = addr & 0xff;
        let high_addr = (addr >> 8) & 0xff;
        let net_id_temp = net_id & 0xff;
        let freq_temp;
        if (freq > 850) {
            freq_temp = freq - 850;
        } else if (freq > 410) {
            freq_temp = freq - 410;
        } else {
            freq_temp = freq; // Or handle error/default case
        }


        let air_speed_temp = this.air_speed_cal(air_speed);
        let buffer_size_temp = this.buffer_size_cal(buffer_size);
        let power_temp = this.power_cal(power);

        let rssi_temp = rssi ? 0x80 : 0x00; // Conditional for RSSI

        let l_crypt = crypt & 0xff;
        let h_crypt = (crypt >> 8) & 0xff;

        this.cfg_reg[3] = high_addr;
        this.cfg_reg[4] = low_addr;
        this.cfg_reg[5] = net_id_temp;
        this.cfg_reg[6] = this.SX126X_UART_BAUDRATE_9600 + air_speed_temp;
        this.cfg_reg[7] = buffer_size_temp + power_temp + 0x20;
        this.cfg_reg[8] = freq_temp;
        this.cfg_reg[9] = 0x03 + rssi_temp;
        this.cfg_reg[10] = h_crypt;
        this.cfg_reg[11] = l_crypt;


        const configBytes = Buffer.from(this.cfg_reg); // Convert array to Buffer

        for (let i = 0; i < 2; i++) {
            this.serialPort.write(configBytes, (err) => { // Send config command
                if (err) {
                    console.log('Serial port write error:', err);
                    return; // Exit if write fails
                }
            });

            await new Promise(resolve => setTimeout(resolve, 200)); // time.sleep(0.2)

            if (this.serialPort.readable) { // Check if data is waiting (using 'readable' event might be better for non-blocking)
                await new Promise(resolve => setTimeout(resolve, 100)); // time.sleep(0.1) - small delay before reading

                const responseBuffer = this.serialPort.read(this.serialPort.bytesToRead); // Read available data
                if (responseBuffer && responseBuffer.length > 0) {
                    if (responseBuffer[0] === 0xC1) {
                        // console.log("parameters setting is :", configBytes); // If you want to log config
                        // console.log("parameters return is  :", responseBuffer); // If you want to log response
                        // Success - you can add more detailed logging if needed
                    } else {
                        // console.log("parameters setting fail :", responseBuffer); // If you want to log failure response
                        // Setting failed - you can add more detailed error handling
                    }
                    break; // Exit loop on response (success or fail)
                }
            } else {
                console.log("trying again!");
                await new Promise(resolve => {
                    this.serialPort.flush(() => { // Flush input buffer before retry
                        console.log('\x1b[1A\r'); // ANSI escape code - might need to handle this differently in Node.js if needed in console
                        if (i === 1) {
                            console.log("setting fail, press Esc to exit and run again");
                            setTimeout(resolve, 2000); // time.sleep(2)
                            console.log('\x1b[1A\r'); // ANSI escape code - might need to handle this differently in Node.js if needed in console
                        } else {
                            resolve();
                        }
                    });
                });

                await new Promise(resolve => setTimeout(resolve, 200)); // time.sleep(0.2)
            }
        }

        


        this.m0Pin.digitalWrite(0); // GPIO.LOW
        this.m1Pin.digitalWrite(0); // GPIO.LOW
        await new Promise(resolve => setTimeout(resolve, 100)); // time.sleep(0.1)
    }

    async receive() {
        // Implement the logic to read data from this.serialPort here
        // and return the received data (or null if no data received).

        // Example (you will need to adapt this based on your needs):
        return new Promise((resolve, reject) => {
            this.serialPort.once('data', (data) => {
                resolve(data.toString('utf8')); // Or process data as needed
            });
            this.serialPort.once('error', (err) => {
                reject(err);
            });
            // Optionally add a timeout to reject the promise if no data received within a certain time
        });
    }
}

// Example of how to use it (you'd need to provide the correct serial port)
async function main() {
    const loraModule = new SX126X("/dev/ttyS0", 868, 0x1234, 22, false); // Example parameters
    // You can add more function calls to loraModule object here to use other functionalities
    console.log("LoRa module initialized and configured.");
}

main();

module.exports = SX126X;